File: targetvideo_flutter_plugin/ios/Classes/TargetvideoFlutterPlugin.swift
import Flutter
import UIKit
import BridSDK

public class TargetvideoFlutterPlugin: NSObject, FlutterPlugin {
    var listOfPlayers: [String: BVPlayer?] = [:]
    private var eventSink: FlutterEventSink?
    private let videoViewFactory: NativeVideoViewFactory
    
    public init(videoViewFactory: NativeVideoViewFactory) {
        self.videoViewFactory = videoViewFactory
        super.init()
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "targetvideo_flutter_plugin", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "targetvideo_flutter_plugin/events", binaryMessenger: registrar.messenger())
        let videoViewFactory = NativeVideoViewFactory()
        let instance = TargetvideoFlutterPlugin(videoViewFactory: videoViewFactory)
        
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.register(videoViewFactory, withId: "targetvideo/player_video_view")
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "load":
            load(call: call, result: result)
        case "pauseVideo":
            if let player = getPlayerWithReference(call: call) {
                player.pause()
            }
            result(nil)
        case "playVideo":
            if let player = getPlayerWithReference(call: call) {
                player.play()
            }
            result(nil)
        case "previous":
            if let player = getPlayerWithReference(call: call) {
                player.previous()
            }
            result(nil)
        case "next":
            if let player = getPlayerWithReference(call: call) {
                player.next()
            }
            result(nil)
        case "mute":
            if let player = getPlayerWithReference(call: call) {
                player.mute()
            }
            result(nil)
        case "unMute":
            if let player = getPlayerWithReference(call: call) {
                player.unmute()
            }
            result(nil)
        case "setFullscreen":
            if let args = call.arguments as? [String: Any],
               let fullscreen = args["fullscreen"] as? Bool,
               let player = getPlayerWithReference(call: call) {
                if fullscreen {
                    player.setFullscreenON()
                } else {
                    player.setFullscreenOFF()
                }
            }
            result(nil)
        case "showControls":
            if let player = getPlayerWithReference(call: call) {
                player.showControls()
            }
            result(nil)
        case "hideControls":
            if let player = getPlayerWithReference(call: call) {
                player.hideControls()
            }
            result(nil)
        case "isAdPlaying":
            if let player = getPlayerWithReference(call: call) {
                let isAdPlaying = player.isAdInProgress()
                result(isAdPlaying)
            } else {
                result(nil)
            }
        case "getPlayerCurrentTime":
            if let player = getPlayerWithReference(call: call) {
                let currentTime = player.getCurrentTime()
                result(currentTime)
            } else {
                result(nil)
            }
        case "getAdDuration":
            if let player = getPlayerWithReference(call: call) {
                let adDuration = player.getAdDuration()
                result(adDuration)
            } else {
                result(nil)
            }
        case "getVideoDuration":
            if let player = getPlayerWithReference(call: call) {
                let duration = player.getDuration()
                result(duration)
            } else {
                result(nil)
            }
        case "isPaused":
            if let player = getPlayerWithReference(call: call) {
                let isPaused = player.isPaused()
                result(isPaused)
            } else {
                result(nil)
            }
        case "isRepeated":
            if let player = getPlayerWithReference(call: call) {
                let isRepeated = player.isRepeated()
                result(isRepeated)
            } else {
                result(nil)
            }
        case "destroyPlayer":
            if let player = getPlayerWithReference(call: call) {
                player.destroy()
            }
            result(nil)
        case "isAutoplay":
            if let player = getPlayerWithReference(call: call) {
                let isAutoplay = player.isAutoplay()
                result(isAutoplay)
            } else {
                result(nil)
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

enum MediaType: String {
    case single = "Single"
    case playlist = "Playlist"
}

// MARK: - Private methods
extension TargetvideoFlutterPlugin {
    
    private func getPlayerReference(call: FlutterMethodCall) -> String? {
        guard let args = call.arguments as? [String: Any],
              let playerReference = args["playerReference"] as? String else { return nil }
        return playerReference
    }
    
    private func getPlayerWithReference(call: FlutterMethodCall) -> BVPlayer? {
        guard let args = call.arguments as? [String: Any],
              let playerReference = args["playerReference"] as? String,
              let player = listOfPlayers[playerReference] else { return nil }
        return player
    }
    
    private func load(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let playerView = UIView()
        guard let args = call.arguments as? [String: Any],
              let playerReference = getPlayerReference(call: call),
              let viewId = args["viewId"] as? Int,
              let view = videoViewFactory.getView(byId: Int64(viewId)) else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments or view not found", details: nil))
            return
        }
        
        let playerId = args["playerId"] as? Int ?? 0
        let mediaId = args["mediaId"] as? Int ?? 0
        let typeOfPlayer = args["typeOfPlayer"] as? String ?? "Single"
        let controlAutoplay = args["controlAutoplay"] as? Bool ?? false
        let scrollOnAd = args["scrollOnAd"] as? Bool ?? false
        let creditsLabelColor = args["creditsLabelColor"] as? String
        let setCornerRadius = args["setCornerRadius"] as? Int ?? 0
        let localization = args["localization"] as? String
        let doubleTapSeek = args["doubleTapSeek"] as? Int ?? 0
        let seekPreview = args["seekPreview"] as? Int ?? 1
        
        if typeOfPlayer == MediaType.single.rawValue {
            loadSingle(view, playerView, playerId, mediaId, playerReference, controlAutoplay, scrollOnAd, creditsLabelColor, setCornerRadius, localization, doubleTapSeek, seekPreview)
        } else if typeOfPlayer == MediaType.playlist.rawValue {
            loadPlaylist(view, playerView, playerId, mediaId, playerReference, controlAutoplay, scrollOnAd, creditsLabelColor, setCornerRadius, localization, doubleTapSeek, seekPreview)
        } else {
            return
        }
    }
    
    private func loadSingle(_ view: UIView,
                            _ playerView: UIView,
                            _ playerId: Int,
                            _ videoId: Int,
                            _ playerReference: String,
                            _ controlAutoplay: Bool,
                            _ scrollOnAd: Bool,
                            _ creditsLabelColor: String?,
                            _ setCornerRadius: Int,
                            _ localization: String?,
                            _ doubleTapSeek: Int,
                            _ seekPreview: Int) {
        
        DispatchQueue.main.async {
            playerView.translatesAutoresizingMaskIntoConstraints = false
            view.addSubview(playerView)
            
            NSLayoutConstraint.activate([
                playerView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
                playerView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
                playerView.topAnchor.constraint(equalTo: view.topAnchor),
                playerView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
            ])
            
            let data = BVData(playerID: Int32(playerId), forVideoID: Int32(videoId))
            let player = BVPlayer(data: data, for: playerView)
            player?.setPlayerReferenceName(playerReference)
            player?.setPlayerLanguage(localization)
            player?.controlAutoplay(controlAutoplay)
            player?.scroll(onAd: scrollOnAd)
            player?.setCornerRadius(Int32(setCornerRadius))
            player?.setSeekSeconds(Int32(doubleTapSeek))
            player?.setSeekPreviewEnabled(Int32(seekPreview))
            if let creditsLabelColor {
                player?.creditsLabelColor(creditsLabelColor)
            }
            
            self.listOfPlayers[playerReference] = player
        }
    }
    
    private func loadPlaylist(_ view: UIView,
                              _ playerView: UIView,
                              _ playerId: Int,
                              _ playlistId: Int,
                              _ playerReference: String,
                              _ controlAutoplay: Bool,
                              _ scrollOnAd: Bool,
                              _ creditsLabelColor: String?,
                              _ setCornerRadius: Int,
                              _ localization: String?,
                              _ doubleTapSeek: Int,
                              _ seekPreview: Int) {
        
        DispatchQueue.main.async {
            playerView.translatesAutoresizingMaskIntoConstraints = false
            view.addSubview(playerView)
            
            NSLayoutConstraint.activate([
                playerView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
                playerView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
                playerView.topAnchor.constraint(equalTo: view.topAnchor),
                playerView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
            ])
            
            let data = BVData(playerID: Int32(playerId), forPlaylistID: Int32(playlistId))
            let player = BVPlayer(data: data, for: playerView)
            player?.setPlayerReferenceName(playerReference)
            player?.setPlayerLanguage(localization)
            player?.controlAutoplay(controlAutoplay)
            player?.scroll(onAd: scrollOnAd)
            player?.setCornerRadius(Int32(setCornerRadius))
            player?.setSeekSeconds(Int32(doubleTapSeek))
            player?.setSeekPreviewEnabled(Int32(seekPreview))
            if let creditsLabelColor {
                player?.creditsLabelColor(creditsLabelColor)
            }
            
            self.listOfPlayers[playerReference] = player
        }
    }
}

// Extension for event streaming
extension TargetvideoFlutterPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        setupEventNetworking()
        setupEventNetworkingForAd()
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
    
    private func setupEventNetworking() {
        NotificationCenter.default.addObserver(self, selector: #selector(eventWriter), name: NSNotification.Name(rawValue: "PlayerEvent"), object: nil)
    }
    
    private func setupEventNetworkingForAd() {
        NotificationCenter.default.addObserver(self, selector: #selector(eventWriter), name: NSNotification.Name(rawValue: "AdEvent"), object: nil)
    }
    
    @objc private func eventWriter(_ notification: NSNotification) {
        var eventDict: [String: Any] = [:]
        
        if let reference = notification.userInfo?["reference"] as? String {
            eventDict["playerReference"] = reference
        }
        
        if notification.name.rawValue == "PlayerEvent" {
            
            if let event = notification.userInfo?["event"] as? String {
                eventDict["type"] = "PlayerEvent"
                eventDict["event"] = event
            }
        }
        
        if notification.name.rawValue == "AdEvent" {
            if let ad = notification.userInfo?["ad"] as? String {
                eventDict["type"] = "AdEvent"
                eventDict["ad"] = ad
            }
        }
        
        if let eventSink = self.eventSink {
            eventSink(eventDict)
        }
    }
}

File: targetvideo_flutter_plugin/android/src/main/kotlin/com/targetvideo/flutter_plugin/TargetvideoFlutterPlugin.kt
package com.targetvideo.flutter_plugin

import android.content.Context
import android.os.Bundle
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel


import tv.brid.sdk.api.BridPlayer
import tv.brid.sdk.api.BridPlayerBuilder
import tv.brid.sdk.player.listeners.BridPlayerListener

/** Main plugin class – bridges Flutter <-> Android. */
class TargetvideoFlutterPlugin : FlutterPlugin,
  MethodChannel.MethodCallHandler,
  EventChannel.StreamHandler {

  private lateinit var methodChannel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private lateinit var viewFactory: NativeVideoViewFactory
  private var eventSink: EventChannel.EventSink? = null

  /** Keeps strong references to all active players by user-defined reference. */
  private val players: MutableMap<String, BridPlayer> = HashMap()

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    val messenger = binding.binaryMessenger
    methodChannel = MethodChannel(messenger, "targetvideo_flutter_plugin").also {
      it.setMethodCallHandler(this)
    }
    eventChannel = EventChannel(messenger, "targetvideo_flutter_plugin/events").also {
      it.setStreamHandler(this)
    }
    viewFactory = NativeVideoViewFactory(messenger)
    binding.platformViewRegistry.registerViewFactory(
      "targetvideo/player_video_view",
      viewFactory
    )
  }

  // ---------- MethodChannel ----------

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) = when (call.method) {
    "load" -> load(call, result)
    "pauseVideo"        -> invokeOnPlayer(call, result) { it.pause() }
    "playVideo"         -> invokeOnPlayer(call, result) { it.play() }
    "previous"          -> invokeOnPlayer(call, result) { it.previous() }
    "next"              -> invokeOnPlayer(call, result) { it.next() }
    "mute"              -> invokeOnPlayer(call, result) { it.setMute(true) }
    "unMute"            -> invokeOnPlayer(call, result) { it.setMute(false) }
    "setFullscreen"     -> setFullscreen(call, result)
    "showControls"      -> invokeOnPlayer(call, result) { it.showControls() }
    "hideControls"      -> invokeOnPlayer(call, result) { it.hideControls() }
    "isAdPlaying"       -> queryPlayer(call, result) { it.isPlayingAd }
    "getPlayerCurrentTime" -> queryPlayer(call, result) { it.currentPosition }
//    "getAdDuration"     -> queryPlayer(call, result) { it.getAdDuration() } - not available on Android
    "getVideoDuration"  -> queryPlayer(call, result) { it.duration }
    "isPaused"          -> queryPlayer(call, result) { it.isPaused }
    "isRepeated"        -> queryPlayer(call, result) { it.isRepeated }
    "destroyPlayer"     -> destroyPlayer(call, result)
    "isAutoplay"        -> queryPlayer(call, result) { it.isAutoplayEnabled }
    else -> result.notImplemented()
  }

  // ---------- EventChannel ----------

  override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
    eventSink = events
  }
  override fun onCancel(arguments: Any?) { eventSink = null }

  // ---------- Private helpers ----------
  private fun load(call: MethodCall, result: MethodChannel.Result) {
    val args = call.arguments as? Map<*, *>
      ?: return result.error("ARG_ERROR", "Arguments must be a map", null)

    // -------------------------------------------------------------------
    // Мandatory – view holder, context, reference IDs
    // -------------------------------------------------------------------
    val viewId = (args["viewId"] as? Number)?.toInt()
      ?: return result.error("ARG_ERROR", "viewId missing or not an int", null)

    val container = viewFactory.getViewHolder(viewId)
      ?: return result.error("VIEW_ERROR", "Native view not found", null)

    val context = container.context

    val playerRef = args["playerReference"] as? String
      ?: return result.error("ARG_ERROR", "playerReference missing", null)

    val playerId = (args["playerId"] as? Number)?.toInt()
      ?: return result.error("ARG_ERROR", "playerId missing or not an int", null)

    // -------------------------------------------------------------------
    // Single  vs  Playlist – један кључ mediaId
    // -------------------------------------------------------------------
    val typeOfPlayer = (args["typeOfPlayer"] as? String)?.ifBlank { "Single" } ?: "Single"

    val mediaId = (args["mediaId"] as? Number)?.toInt()
      ?: return result.error("ARG_ERROR", "mediaId missing or not an int", null)

    // -------------------------------------------------------------------
    // Optional parameters  (safe–cast + подразумеване вредности)
    // -------------------------------------------------------------------
    val localization    = args["localization"]     as? String  ?: "en"
    val controlAutoplay = args["controlAutoplay"]  as? Boolean ?: false
    val scrollOnAd      = args["scrollOnAd"]       as? Boolean ?: false
    val creditsColor    = args["creditsLabelColor"] as? String
    val cornerRadius    = (args["setCornerRadius"] as? Number)?.toInt() ?: 0
    val doubleTapSeek   = (args["doubleTapSeek"]   as? Number)?.toInt() ?: 0
    val seekPreview     = (args["seekPreview"]     as? Number)?.toInt() ?: 1

    // -------------------------------------------------------------------
    // Build Brid Player
    // -------------------------------------------------------------------
    val bridPlayer = BridPlayerBuilder(context, container)
      .fullscreen(false)
      .setPlayerLanguage(localization)
      .setCornerRadius(cornerRadius)
      .setSeekSeconds(doubleTapSeek)
      .useVpaidSupport(false)
      .build().apply {
        setPlayerReference(playerRef)
        enableAutoplay(controlAutoplay)
        setCreditsLabelColor(creditsColor)
        enableAdControls(!scrollOnAd)
        setSeekPreview(seekPreview)
        setBridListener(createListener())
      }

    // -------------------------------------------------------------------
    // Load content — исти mediaId, различит Brid API
    // -------------------------------------------------------------------
    if (typeOfPlayer.equals("Playlist", ignoreCase = true))
      bridPlayer.loadPlaylist(playerId, mediaId)  // mediaId = playlistId
    else
      bridPlayer.loadVideo(playerId, mediaId)     // mediaId = videoId

    // Cache reference & return
    players[playerRef] = bridPlayer
    result.success(null)
  }



  private fun setFullscreen(call: MethodCall, result: MethodChannel.Result) =
    invokeOnPlayer(call, result) { player ->
      val fullscreen = (call.argument<Boolean>("fullscreen") ?: false)
      player.setFullScreen(fullscreen)
    }

  private fun destroyPlayer(call: MethodCall, result: MethodChannel.Result) {
    getPlayer(call)?.apply {
      release()
      players.remove(call.argument<String>("playerReference"))
    }
    result.success(null)
  }

  /** Executes [action] on the player and returns success(null). */
  private inline fun invokeOnPlayer(
    call: MethodCall,
    result: MethodChannel.Result,
    crossinline action: (BridPlayer) -> Unit
  ) {
    getPlayer(call)?.let { action(it) }
    result.success(null)
  }

  /** Executes [query] and returns its value through [result]. */
  private inline fun <T> queryPlayer(
    call: MethodCall,
    result: MethodChannel.Result,
    crossinline query: (BridPlayer) -> T
  ) {
    result.success(getPlayer(call)?.let { query(it) })
  }

  private fun getPlayer(call: MethodCall): BridPlayer? =
    players[call.argument<String>("playerReference")]

  /** Generates listener that forwards both player and ad events to Flutter. */
  private fun BridPlayer.createListener(): BridPlayerListener = BridPlayerListener { status, ref ->
    val type = if (isAdEvent(status)) "AdEvent" else "PlayerEvent"
    val map  = HashMap<String, Any>().apply {
      put("playerReference", ref ?: "")
      put("type", type)
      if (type == "AdEvent") put("ad", status) else put("event", status)
    }
    eventSink?.success(map)
    Log.d("BridPlayerEvent", "$type -> $status ($ref)")
  }

  /** Simple heuristic – all ad event codes start with "ad" or are in known set. */
  private fun isAdEvent(code: String): Boolean =
    code.startsWith("ad") || code in setOf("r","i","v","s","ae","adbstart")

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
    players.values.forEach { it.release() }
    players.clear()
  }
}
